src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,455 @@
+	buffer->size = size;
+	buffer->w_ptr = buffer->addr;
+	buffer->r_ptr = buffer->addr;
+	buffer->end_addr = buffer->addr + buffer->size;
+	buffer->avail = 0;
+
+	return 0;
+}
+
+void lib_manager_dma_buffer_free(struct lib_manager_dma_buf *buffer)
+{
+	rfree((void *)buffer->addr);
+	memset(buffer, 0, sizeof(struct lib_manager_dma_buf));
+}
+
+/**
+ * \brief Start, init and alloc DMA and buffer used by loader.
+ *
+ * \return 0 on success, error code otherwise.
+ */
+static int lib_manager_dma_init(struct lib_manager_dma_ext *dma_ext, uint32_t dma_id)
+{
+	uint32_t addr_align;
+	int ret = -ENOMEM;
+
+	/* request DMA in the dir HMEM->LMEM */
+	dma_ext->dma.dmac = dma_get(DMA_DIR_HMEM_TO_LMEM, 0, DMA_DEV_HOST,
+			DMA_ACCESS_EXCLUSIVE);
+	if (!dma_ext->dma.dmac) {
+		tr_err(&lib_manager_tr, "dma_ext_init(): dma.dmac = NULL");
+		return -ENODEV;
+	}
+
+	/* get required address alignment for dma buffer */
+	ret = dma_get_attribute(dma_ext->dma.dmac, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
+					&addr_align);
+	if (ret < 0) {
+		tr_err(&lib_manager_tr, "dma_get_attribute(): ret = %d", ret);
+		return ret;
+	}
+
+	dma_ext->chan = dma_channel_get(dma_ext->dma.dmac, dma_id);
+	if (!dma_ext->chan) {
+		tr_err(&lib_manager_tr, "dma_channel_get(): dma_ext->chan = NULL");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * \brief Stop, deinit and free DMA and buffer used by loader.
+ *
+ * \return 0 on success, error code otherwise.
+ */
+
+static int lib_manager_dma_deinit(struct lib_manager_dma_ext *dma_ext)
+{
+	dma_channel_put(dma_ext->chan);
+	dma_put(dma_ext->dma.dmac);
+	return 0;
+}
+
+static int lib_manager_load_data_from_host(struct lib_manager_dma_ext *dma_ext, uint32_t size)
+{
+	struct dma_sg_config config;
+	uint32_t elem_addr;
+	uint32_t elem_num, elem_size;
+	int ret = -ENOMEM;
+	uint32_t free_bytes = 0;
+	uint32_t avail_bytes = 0;
+
+	elem_addr = (uint32_t)dma_ext->dmabp.addr;
+
+	elem_size = sizeof(uint64_t) * DMA_ELEM_SIZE;
+	elem_num = size / elem_size;
+
+	config.direction = DMA_DIR_HMEM_TO_LMEM;
+	config.src_width = sizeof(uint32_t);
+	config.dest_width = sizeof(uint32_t);
+	config.cyclic = 0;
+
+	ret = dma_sg_alloc(&config.elem_array, SOF_MEM_ZONE_RUNTIME,
+			   config.direction, elem_num, elem_size, elem_addr, 0);
+
+	if (ret < 0)
+		return ret;
+
+	ret = dma_set_config(dma_ext->chan, &config);
+	if (ret < 0)
+		return ret;
+
+	ret = dma_start(dma_ext->chan);
+	if (ret < 0)
+		return ret;
+
+	/* Wait till whole data acquired */
+	while (avail_bytes < size) {
+		/* get data sizes from DMA */
+		ret = dma_get_data_size(dma_ext->chan, &avail_bytes, &free_bytes);
+		if (ret < 0)
+			return ret;
+
+		/* Exit if avail_bytes == 0 */
+		if (avail_bytes == 0)
+			break;
+#ifndef __ZEPHYR__
+		wait_delay(1);
+#else
+		Z_TIMEOUT_CYC(1);
+#endif
+		if (ret < 0)
+			return ret;
+	}
+
+	ret = dma_stop_delayed(dma_ext->chan);
+	if (ret < 0)
+		return ret;
+
+	dma_sg_free(&config.elem_array);
+
+	dcache_invalidate_region((void *)dma_ext->dmabp.addr, size);
```

Comment:
why is this here? Have we just used DMA to copy data into that area? If so, them it's too late to invalidate caches here. If they really have to be invalidated, then before the DMA transfer. If you end up using this and moving it elsewhere, please add the sparse annotations as explained above

Reply:
We just used DMA to copy new data to the buffer, but before accessing them we need invalidate cache. Otherwise we could read old data from cache instead of the new one just transferred using DMA.

Reply:
if we had data in cache for those addresses, while DMA was writing there the CPU could speculatively write caches back to RAM too. Therefore caches should be invalidated before letting DMA write to that memody

Reply:
cache is used only once during loading then modules are moved from cache to physical memory.